### PR TITLE
Add examples for `functional_serializers` in API docs

### PR DESCRIPTION
**Pull request by hungtsetse**
## Change Summary
Add examples to `functional_serializers` in API docs.

## Related issue number
#8234 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @hramezani

### CODE REVIEW

**Comment by hungtsetse** [comment]
Please review

File: pydantic/functional_serializers.py [hunk 1]
```diff
@@ -18,6 +18,28 @@
 class PlainSerializer:
     """Plain serializers use a function to modify the output of serialization.
 
+    If you want to customize the serialization for annotated types.
```
Comment (sydney-runkle) [inline pydantic/functional_serializers.py:21]:
```suggestion
    This is particularly helpful when you want to customize the serialization for annotated types.
```

File: pydantic/functional_serializers.py [hunk 2]
```diff
@@ -18,6 +18,28 @@
...
+    CustomStr = Annotated[
+        List, PlainSerializer(lambda x: ' '.join(x), return_type=str)
+    ]
+
+    class StudentModel(BaseModel):
+        courses: CustomStr
+
+    student = StudentModel(courses=['Math', 'Chemistry', 'English'])
+    print(student.model_dump())
+    #> {'courses': 'Math Chemistry English'}
```
Comment (sydney-runkle) [inline pydantic/functional_serializers.py:40]: Great example!

File: pydantic/functional_serializers.py [hunk 3]
```diff
@@ -151,6 +173,26 @@ def field_serializer(
...
+    from typing import Set
+
+    from pydantic import BaseModel, field_serializer
+
+    class StudentModel(BaseModel):
+        name: str = 'Jane'
+        courses: Set[str]
+
+        @field_serializer('courses', when_used='json')
+        def serialize_course_in_order(courses: Set[str]):
```
Comment (sydney-runkle) [inline pydantic/functional_serializers.py:188]:
```suggestion
        def serialize_courses_in_order(courses: Set[str]):
```

File: pydantic/functional_serializers.py [hunk 4]
```diff
@@ -151,6 +173,26 @@ def field_serializer(
 ) -> Callable[[Any], Any]:
     """Decorator that enables custom field serialization.
 
+    For example, a field of `set` type to mitigate duplication. We want this field to be serialized as a sorted list.
```
Comment (sydney-runkle) [inline pydantic/functional_serializers.py:176]:
```suggestion
    In the below example, a field of type `set` is used to mitigate duplication. A `field_serializer` is used to serialize the data as a sorted list.
```

File: pydantic/functional_serializers.py [hunk 5]
```diff
@@ -217,6 +259,30 @@ def model_serializer(
...
+        def serialize_model(self):
+            return {'name': self.name.title(), 'courses': sorted(self.courses)}
+
+    student = StudentModel(
+        name='jane doe', courses={'Math', 'Chemistry', 'English'}
+    )
+    print(student.model_dump_json())
+    #> {"name":"Jane Doe","courses":["Chemistry","English","Math"]}
+    ```
+
```
Comment (sydney-runkle) [inline pydantic/functional_serializers.py:285]: I like the thought here, but I think we probably want to use an example that couldn't be achieved by using 2 different `field_serializer` decorators.

For example, these examples from the usage docs are more `model_serializer` specific (there's one relevant example per section linked below)
* https://docs.pydantic.dev/latest/concepts/serialization/#overriding-the-return-type-when-dumping-a-model
* https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
@hungtsetse,

Thanks for your contribution! Looking great overall. I've left a bit of feedback / a few change requests.

Could you please also add examples for `WrapSerializer`? Thanks for your work on this!

**Comment by sydney-runkle** [comment]
@hungtsetse,

Sure thing! If you're able to make the changes soon, we can get these into the upcoming release!

File: pydantic/functional_serializers.py [hunk 6]
```diff
@@ -217,6 +259,30 @@ def model_serializer(
...
+        @model_serializer()
+        def serialize_model(self):
+            if self.unit == 'F':
+                return {'unit': 'C', 'value': int((self.value - 32) / 1.8)}
+            return {'unit': self.unit, 'value': self.value}
+
+    temperature = TemperatureModel(unit='F', value=212)
+    print(temperature.model_dump())
+    #> {'unit': 'C', 'value': 100}
+    ```
```
Comment (sydney-runkle) [inline pydantic/functional_serializers.py:284]: Perfect example! Love this change.

File: pydantic/functional_serializers.py (hunk 5 above)
Comment (hungtsetse) [reply pydantic/functional_serializers.py]: I've rewrited it with another example.

https://github.com/pydantic/pydantic/blob/fb0b40a45cc1d75318e4a809eb72c5460e8fe5cd/pydantic/functional_serializers.py#L264-L284

File: pydantic/functional_serializers.py (hunk 6 above)
Comment (hungtsetse) [reply pydantic/functional_serializers.py:284]: Glad to hear that.
The remaining is `WrapSerializer` which iam working on.

Comment (sydney-runkle) [reply pydantic/functional_serializers.py:284]: Fantastic!

**Comment by hungtsetse** [comment]
@sydney-runkle 

I tried to come up with an example of `WrapSerializer` which can fully demonstrate usage of `handler` parameter in the wrappered serializer function.
I've gone through the [https://docs.pydantic.dev/latest/concepts/serialization](https://docs.pydantic.dev/latest/concepts/serialization/#overriding-the-return-type-when-dumping-a-model), but I'm still not very sure if it's a good example.
https://github.com/pydantic/pydantic/blob/1366fc636658825f583be05d81cacd784a85e580/pydantic/functional_serializers.py#L96-L104

Thank you in advance for your review.

**Comment by sydney-runkle** [comment]
@hungtsetse,

> I've gone through the [https://docs.pydantic.dev/latest/concepts/serialization](https://docs.pydantic.dev/latest/concepts/serialization/#overriding-the-return-type-when-dumping-a-model), but I'm still not very sure if it's a good example.

Thanks for your work on this - I think you're off to a good start. Indeed, wrap serializers can be a big confusing. I think that the example we're discussing in this issue does a good job showcasing how it can be useful to perform standard serialization first, then add your own custom logic. Specifically, see my comment at the end here: https://github.com/pydantic/pydantic/issues/8536#issuecomment-1892506726.

Feel free to steal that example, or use a similar one. The main thing is that you call `handler(source)`, then run your own logic.

Looking forward to a last review - thanks again for your help on this!

**Comment by hungtsetse** [comment]
@sydney-runkle 

Thanks for sharing the code snippet #8536, I found it's extremely helpful.
I added the final example, it is good for review now.

File: pydantic/functional_serializers.py [hunk 7]
```diff
@@ -61,6 +83,59 @@ class WrapSerializer:
     """Wrap serializers receive the raw inputs along with a handler function that applies the standard serialization
     logic, and can modify the resulting value before returning it as the final output of serialization.
 
+    A scenario in which a model's subclass containing datetime fields is transformed to UTC without modifying subclass.
```
Comment (sydney-runkle) [inline pydantic/functional_serializers.py:86]:
```suggestion
    For example, here's a scenario in which a wrap serializer transforms timezones to UTC **and** utilizes the existing `datetime` serialization logic.
```

**Review by sydney-runkle — APPROVED** [review]
Looks great!

The example that you used works nicely with this one: https://docs.pydantic.dev/latest/why/#customisation and this one: https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers, so I think it fits well.

**Comment by sydney-runkle** [comment]
@hungtsetse,

Awesome job on these examples. Will go ahead and merge this!

If you are interested in contributing further, we have a variety of other issues marked as `help wanted` or `good first issue` :).

**Pull request merged**